tacacs_plus/async_client.py

```diff
@@ -0,0 +1,399 @@
+            error_message = resp_header.__class__.__name__ + '\n'
+            error_message += 'recv header <%s>' % resp_header + '\n'
+            error_message += str(resp_header.packed)
+            logger.error(error_message)
+            raise socket.error
+
+        # read the number of bytes specified in the response header
+        body_bytes = b''
+        remaining = resp_header.length
+        while remaining > 0:
+            body_bytes += await reader.read(remaining)
+            remaining = resp_header.length - len(body_bytes)
+        return TACACSPacket(resp_header, body_bytes, self.secret)
+
+    async def authenticate(
+        self,
+        username,
+        password,
+        priv_lvl=TAC_PLUS_PRIV_LVL_MIN,
+        authen_type=TAC_PLUS_AUTHEN_TYPE_ASCII,
+        chap_ppp_id=None,
+        chap_challenge=None,
+        rem_addr=TAC_PLUS_VIRTUAL_REM_ADDR,
+        port=TAC_PLUS_VIRTUAL_PORT,
+        reader=None,
+        writer=None
+    ):
+        """
+        Authenticate to a TACACS+ server with a username and password.
+
+        :param username:
+        :param password:
+        :param priv_lvl:
+        :param authen_type:    TAC_PLUS_AUTHEN_TYPE_ASCII,
+                               TAC_PLUS_AUTHEN_TYPE_PAP,
+                               TAC_PLUS_AUTHEN_TYPE_CHAP
+        :param chap_ppp_id:    PPP ID when authen_type == 'chap'
+        :param chap_challenge: challenge value when authen_type == 'chap'
+        :param rem_addr:       AAA request source, default to TAC_PLUS_VIRTUAL_REM_ADDR
+        :param port:           AAA port, default to TAC_PLUS_VIRTUAL_PORT
+        :param reader:         asyncio.StreamReader instance
+        :param writer:         asyncio.StreamWriter instance
+        :return:               TACACSAuthenticationReply
+        :raises:               socket.timeout, socket.error
+        """
+        start_data = b''
+        if authen_type in (TAC_PLUS_AUTHEN_TYPE_PAP,
+                           TAC_PLUS_AUTHEN_TYPE_CHAP):
+            self.version_min = TAC_PLUS_MINOR_VER_ONE
+
+            if authen_type == TAC_PLUS_AUTHEN_TYPE_PAP:
+                start_data = password.encode()
+
+            if authen_type == TAC_PLUS_AUTHEN_TYPE_CHAP:
+                if not isinstance(chap_ppp_id, str):
+                    raise ValueError('chap_ppp_id must be a string')
+                if len(chap_ppp_id) != 1:
+                    raise ValueError('chap_ppp_id must be a 1-byte string')
+                if not isinstance(chap_challenge, str):
+                    raise ValueError('chap_challenge must be a string')
+                if len(chap_challenge) > 255:
+                    raise ValueError('chap_challenge may not be more 255 bytes')
+
+                start_data = chap_ppp_id.encode()
+                start_data += chap_challenge.encode()
+                data_to_md5 = (chap_ppp_id + password + chap_challenge).encode()
+                start_data += md5(data_to_md5).digest()
```

Comment:
Nit: this is not a blocker, but I am wondering if we can extract this logic since it is basically the same in the sync and async client..

Reply:
Not sure how to reuse this code.

Suggestion 1: extract `__init__` and `version` methods from Client/AsyncClient to new BaseClient class, and add `staticmethod` to deal with authen data packing. I do not like this variant.

Suggestion 2: move code to new function in authentication.py, as it is related to authentication, then patch both sync and async clients.
Function signature be like `(password, authen_type, chap_ppp_id, chap_challenge) -> bytes`

Reply:
```
def make_start_data(password, authen_type, chap_ppp_id, chap_challenge):
    if authen_type == TAC_PLUS_AUTHEN_TYPE_ASCII:
        return six.b('')
    elif authen_type == TAC_PLUS_AUTHEN_TYPE_PAP:
        return six.b(password)
    elif authen_type == TAC_PLUS_AUTHEN_TYPE_CHAP:
        if not isinstance(chap_ppp_id, six.string_types):
            raise ValueError('chap_ppp_id must be a string')
        if len(chap_ppp_id) != 1:
            raise ValueError('chap_ppp_id must be a 1-byte string')
        if not isinstance(chap_challenge, six.string_types):
            raise ValueError('chap_challenge must be a string')
        if len(chap_challenge) > 255:
            raise ValueError('chap_challenge may not be more 255 bytes')
        return (
                six.b(chap_ppp_id) +
                six.b(chap_challenge) +
                md5(six.b(
                    chap_ppp_id + password + chap_challenge
                )).digest()
        )
```

Reply:
I think a function will do without having to bring in a base class.. Anyway, I would defer this change to some refactoring I would like to do after this PR lands..